### ISSUE TITLE
log and log10 values showing amazing performance.

### ISSUE THREAD

**Issue by abhimanyu3** [facebook/prophet#1387]
I am detecting anomaly in realtime and we received data every micro and nano second so I am grouping by the sum every minute and trying to detect anomaly with the below code:-

My data looks like below :- 
![Screen Shot 2020-03-11 at 11 37 03 AM](https://user-images.githubusercontent.com/8762047/76434878-a8653e00-638c-11ea-954b-126d0d0898fc.png)

```
def fit_predict_model(dataframe, interval_width = 0.99, changepoint_range = 0.8):
    m = Prophet(daily_seasonality = True, yearly_seasonality = False, weekly_seasonality = True,
                seasonality_mode = 'additive', 
                interval_width = interval_width,
                changepoint_range = changepoint_range)
    m = m.fit(dataframe)
    forecast = m.predict(dataframe)
    forecast['fact'] = dataframe['y'].reset_index(drop = True)
    return forecast
    
pred = fit_predict_model(df)
```

```
def detect_anomalies(forecast):
    forecasted = forecast[['ds','trend', 'yhat', 'yhat_lower', 'yhat_upper', 'fact']].copy()
    #forecast['fact'] = df['y']

    forecasted['anomaly'] = 0
    forecasted.loc[forecasted['fact'] > forecasted['yhat_upper'], 'anomaly'] = 1
    forecasted.loc[forecasted['fact'] < forecasted['yhat_lower'], 'anomaly'] = -1

    #anomaly importances
    forecasted['importance'] = 0
    forecasted.loc[forecasted['anomaly'] ==1, 'importance'] = \
        (forecasted['fact'] - forecasted['yhat_upper'])/forecast['fact']
    forecasted.loc[forecasted['anomaly'] ==-1, 'importance'] = \
        (forecasted['yhat_lower'] - forecasted['fact'])/forecast['fact']
    
    return forecasted

pred = detect_anomalies(pred)
```

**1.) Without Log**
When I am not using log my r2 is pretty amazing but MSE and MAE are pretty bad and my model is overfitting:-

R2 :- 0.983552606580086
MSE :- 4124213894.9169807
MAE :- 48678.028212689096

**2.) When using log** 
When I used **log** in my model performed really well though the value of R2 goes a little down but MSE and MAE are amazing.

R2 :- 0.9753557883665813
MSE :- 0.0035934341302893847
MAE :- 0.04046356773306616

**3.) When I used log10** 

When I used **log10** my model performance was even better.

R2 :- 0.9752638873073614
MSE :- 0.0006801307000479876
MAE :- 0.01763236820020733

**I am taking log because standard deviation is very high in my data and there is no skewness. 
The reason is standard deviation.**
Its too good to be true. Am I missing something here ?

**Comment by benmwhite** [comment]
You can't compare the MSE and MAE across these runs because they're all in different units thanks to the transformations, and [R^2 isn't an appropriate error measurement in time series](https://stats.stackexchange.com/questions/101546/what-is-the-problem-with-using-r-squared-in-time-series-models) because the observations are time-dependent and not independent.

I'd try evaluating a metric like MAPE instead, over a holdout set of future data points (check out prophet's [built-in cross-validation](https://facebook.github.io/prophet/docs/diagnostics.html) functionality too). Since everything will be expressed in percentage error you'll actually be able to make comparisons between the three models.

All that said, if the log transformed models are still actually beating the original model in MAPE on your validation sets then that probably indicates that your daily seasonality effect is actually [multiplicative rather than additive](https://facebook.github.io/prophet/docs/multiplicative_seasonality.html) (see [this related issue discussion](https://github.com/facebook/prophet/issues/1363)).

**Comment by abhimanyu3** [comment]
Thanks for your feedback:-

My daily seasonality is defiantly multiplicative:-
![Screen Shot 2020-03-11 at 1 30 06 PM](https://user-images.githubusercontent.com/8762047/76446094-e5393100-639c-11ea-842c-16b6d8bedc2c.png)

But weekly seasonality is additive :- 
![Screen Shot 2020-03-11 at 1 32 46 PM](https://user-images.githubusercontent.com/8762047/76446120-eff3c600-639c-11ea-85e6-0ff609361346.png)

Actually I have 50 different domains and have almost the same pattern. 
So, in that case shall I put daily seasonality as false.

**Comment by abhimanyu3** [comment]
Also if I am predicting the next 15 minutes can you please help me in putting the parameters :-  
I got hard time understating initial period and horizon.

Basically I am finding anomalies in real time its whether I do it without prediction or with prediction doesn't matter.
from fbprophet.diagnostics import cross_validation
df_cv = cross_validation(m, initial='730 days', period='180 days', horizon = '365 days')
df_cv.head()

**Comment by abhimanyu3** [comment]
I calculated the MAPE and without taking the log its 
MAE 22442.64030820021
MAPE 1.4049310567858067

with the log its 
MAE 0.03244082948430389
MAPE 0.22710828320089033


I took last 10 days of data and predicted the next 30 minutes:-

original df shape 
14400, 2

```
prediction_size = 30
train_df = df[:-prediction_size]
train_df.tail(3)
m = Prophet(daily_seasonality = True, yearly_seasonality = False, weekly_seasonality = True,
                seasonality_mode = 'additive', 
                interval_width = 0.95,
                changepoint_range = 0.8)
m.fit(train_df);
future = m.make_future_dataframe(periods=prediction_size, freq='min')

forecast = m.predict(future)

def make_comparison_dataframe(historical, forecast):
    """Join the history with the forecast.
    
       The resulting dataset will contain columns 'yhat', 'yhat_lower', 'yhat_upper' and 'y'.
    """
    return forecast.set_index('ds')[['yhat', 'yhat_lower', 'yhat_upper']].join(historical.set_index('ds'))

cmp_df = make_comparison_dataframe(df, forecast)

def calculate_forecast_errors(df, prediction_size):
    """Calculate MAPE and MAE of the forecast.
    
       Args:
           df: joined dataset with 'y' and 'yhat' columns.
           prediction_size: number of days at the end to predict.
    """
    
    # Make a copy
    df = df.copy()
    
    # Now we calculate the values of e_i and p_i according to the formulas given in the article above.
    df['e'] = df['y'] - df['yhat']
    df['p'] = 100 * df['e'] / df['y']
    
    # Recall that we held out the values of the last `prediction_size` days
    # in order to predict them and measure the quality of the model. 
    
    # Now cut out the part of the data which we made our prediction for.
    predicted_part = df[-prediction_size:]
    
    # Define the function that averages absolute error values over the predicted part.
    error_mean = lambda error_name: np.mean(np.abs(predicted_part[error_name]))
    
    # Now we can calculate MAPE and MAE and return the resulting dictionary of errors.
    return {'MAPE': error_mean('p'), 'MAE': error_mean('e')}

for err_name, err_value in calculate_forecast_errors(cmp_df, prediction_size).items():
    print(err_name, err_value)
```

Reference:-
(https://medium.com/open-machine-learning-course/open-machine-learning-course-topic-9-part-3-predicting-the-future-with-facebook-prophet-3f3af145cdc)

@benmwhite

**Comment by benmwhite** [comment]
Seems like log transforms are the way to go, as best as I can tell. If you're going to be doing anomaly detection based on predicted values then just make sure that your future horizon time period makes sense for how often you're going to be refreshing the model, and check the predicted uncertainty bands visually to make sure they're not too wide or narrow for your use case.

**Comment by hmamirchishti** [comment]
Hey guys, I have a question. In this problem you are not coming back from the log domain as in anomaly detection is does make sense but what if we are predicting sales? In that case, we have to report actual sales.

The problem I'm dealing with is when I take antilog it shoot the error even when the model has been trained optimally, because the difference between antilog(12) and antilog(12.1) is very big, considering 12 and 12.1 are actual and forecast values respectively.
